scripts/kconfig/merge_config.py

```diff
@@ -3,50 +3,91 @@
-    parser.add_argument('-m', dest='runmake', action='store_false',
-        help='only merge the fragments, do not execute the make command')
-    parser.add_argument('-n', dest='alltarget', action='store_const', const='allnoconfig', default='alldefconfig',
-        help='use allnoconfig instead of alldefconfig')
     parser.add_argument('-r', dest='warnredun', action='store_true',
-        help='list redundant entries when merging fragments')
+                        help='list redundant entries when merging fragments')
     parser.add_argument('-O', dest='output_dir',
-        help='to put generated output files', default='.')
+                        help='to put generated output files', default='.')
     parser.add_argument('config', nargs='+')
 
     return parser.parse_args()
 
 
+def get_config_name(line):
+    # '# CONFIG_FOO is not set' should be treated by merge_config as a
+    # state like any other
+
+    is_not_set_pattern = re.compile('^# (CONFIG_[a-zA-Z0-9_]*) is not set.*')
+    pattern            = re.compile('^(CONFIG_[a-zA-Z0-9_]*)[= ].*')
+
+    match            =            pattern.match(line)
+    match_is_not_set = is_not_set_pattern.match(line)
+
+    if match_is_not_set:
+        return match_is_not_set.group(1)
+    elif match:
+        return match.group(1)
+
+    return "" # No match
+
 def main():
     args = arguments_parse()
 
-    pattern = re.compile('^(CONFIG_[a-zA-Z0-9_]*)[= ].*')
-    config_values = {}
-
-    for config_file in args.config[:]:
-        print("Merging %s" % config_file)
-        with open(config_file, 'r') as file:
-            for line in file:
-                match = pattern.match(line)
-                if match:
-                    config_name = match.group(1)
-                    config_full = match.group(0)
-                    if config_name in config_values:
-                        if config_values[config_name] != config_full and not args.quiet:
-                            print("Value of %s is redefined by fragment %s" % (config_name, config_file))
-                            print("Previous value %s" % config_values[config_name])
-                            print("new value %s" % config_full)
-                        elif args.warnredun:
-                            print("Value of %s is redundant by fragment %s" % (config_name, config_file))
-                    config_values[config_name] = config_full
-
-    if args.runmake:
-        print("Running make not yet supported")
-
-    with open('%s/.config' % args.output_dir, 'w') as out_config:
-        out_config.write('\n'.join(map(lambda x: x[1], config_values.items())))
+    # Create a datastructure in 'conf' that looks like this:
+    # [
+    #   ("CONFIG_UART", "CONFIG_UART=y"),
+    #   ("CONFIG_ARM", "# CONFIG_ARM is not set")
+    # ]
+    # In other words [(config_name, config_line), ... ]
+    #
+    # Note that "# CONFIG_ARM is not set" is not the same as a
+    # comment, it has meaning (for now).
```

Comment:
I disagree. IMHO "# is not set" should eventually be deprecated in favour of "=n" for obvious reasons.

Zephyr usability trumps Linux kernel Kconfig compatibility.

Reply:
> I disagree. IMHO "# is not set" should eventually be deprecated in favour of "=n" for obvious reasons.

I can see why you say so.

There are many more people who are familiar with the Linux style than the Zephyr style, though, and I think it's "obvious" that we want to obey principle of least surprise, especially since Zephyr is a Linux Foundation project.

@nashif, what's your decision?

Reply:
If I may weigh in, I think that both formats should be supported and should do exactly the same:
```
CONFIG_BT=n
# CONFIG_BT is not set
```
Should both turn Bluetooth OFF if present in the .conf or defconfig files

Reply:
Maybe I've missed something, but does `conf` treat those 2 differently in the `.config` file?

Reply:
To get this merged, I suggest we retain KBuild compatibilty for now, and then we can discuss the long-term Kconfig syntax at a later point.